syntax = "proto2";

message StructuredMessage {
    required BotState BotState = 1;
    required Player Player = 2;
    optional Target Target = 3;
}

enum BotState {
    Stopped = 0;
    Running = 1;
    DumpPos = 2;
}

message Player {
    required Unit UnitInfo = 1;
    required Position PosInfo = 2;
    // Since booleans don't work in lua-pb, we pack them into a single u64 instead
    // The flags in order from least significant bit to most significant bit are:
    // IsOutdoors, IsMounted, IsFlying
    required uint64 Flags = 3;
}

message Target {
    required Unit UnitInfo = 1;
}

enum Class {
    None = 0;
    Warrior = 1;
    Paladin = 2;
    Hunter = 3;
    Rogue = 4;
    Priest = 5;
    DeathKnight = 6;
    Shaman = 7;
    Mage = 8;
    Warlock = 9;
    Monk = 10;
    Druid = 11;
}

message Unit {
    required string Name = 1;
    required int64 Level = 2;
    optional Class Class = 3;
    required int64 HealthCurrent = 4;
    required int64 HealthMax = 5;
    optional int64 PowerCurrent = 6;
    optional int64 PowerMax = 7;
}

message Position {
    required uint64 MapX = 1;
    required uint64 MapY = 2;
    required sint64 WorldX = 3;
    required sint64 WorldY = 4;
    required uint64 InstanceId = 5;
    required uint64 Facing = 6;
}
